src/main/java/org/jboss/dmr/stream/ModelReaderImpl.java

```diff
@@ -0,0 +1,761 @@
+import static org.jboss.dmr.stream.ModelConstants.QUOTE;
+import static org.jboss.dmr.stream.Utils.isNumberChar;
+import static org.jboss.dmr.stream.Utils.isHexNumberChar;
+import static org.jboss.dmr.stream.Utils.isWhitespace;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.Reader;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+/**
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ */
+final class ModelReaderImpl implements ModelReader {
+
+    private static final char[] BIG = ModelConstants.BIG.toCharArray();
+    private static final char[] BYTES = ModelConstants.BYTES.toCharArray();
+    private static final char[] DECIMAL = ModelConstants.DECIMAL.toCharArray();
+    private static final char[] EXPRESSION = ModelConstants.EXPRESSION.toCharArray();
+    private static final char[] FALSE = ModelConstants.FALSE.toCharArray();
+    private static final char[] INFINITY = ModelConstants.INFINITY.toCharArray();
+    private static final char[] INTEGER = ModelConstants.INTEGER.toCharArray();
+    private static final char[] NAN = ModelConstants.NAN.toCharArray();
+    private static final char[] TRUE = ModelConstants.TRUE.toCharArray();
+    private static final char[] TYPE_BIG_DECIMAL = ModelType.BIG_DECIMAL.toString().toCharArray();
+    private static final char[] TYPE_BIG_INTEGER = ModelType.BIG_INTEGER.toString().toCharArray();
+    private static final char[] TYPE_BOOLEAN = ModelType.BOOLEAN.toString().toCharArray();
+    private static final char[] TYPE_BYTES = ModelType.BYTES.toString().toCharArray();
+    private static final char[] TYPE_DOUBLE = ModelType.DOUBLE.toString().toCharArray();
+    private static final char[] TYPE_INT = ModelType.INT.toString().toCharArray();
+    private static final char[] TYPE_LIST = ModelType.LIST.toString().toCharArray();
+    private static final char[] TYPE_LONG = ModelType.LONG.toString().toCharArray();
+    private static final char[] TYPE_EXPRESSION = ModelType.EXPRESSION.toString().toCharArray();
+    private static final char[] TYPE_OBJECT = ModelType.OBJECT.toString().toCharArray();
+    private static final char[] TYPE_PROPERTY = ModelType.PROPERTY.toString().toCharArray();
+    private static final char[] TYPE_STRING = ModelType.STRING.toString().toCharArray();
+    private static final char[] TYPE_TYPE = ModelType.TYPE.toString().toCharArray();
+    private static final char[] TYPE_UNDEFINED = ModelType.UNDEFINED.toString().toCharArray();
+    private static final char[] UNDEFINED = ModelConstants.UNDEFINED.toCharArray();
+    private final Reader in;
+    private final ModelGrammarAnalyzer analyzer;
+    private char[] buffer = new char[ 1024 ];
+    private int position;
+    private int limit;
+    private int numberOffset;
+    private int numberLength;
+    private int stringOffset;
+    private int stringLength;
+    private byte[] bytesValue;
+    private ModelType typeValue;
+    private Number numberValue;
```

Comment:
Using separate fields for the int or long value may reduce allocation churn a bit while reading the model, by not maintaining boxed Integer and Long instances.